feat: Allow to pass custom data on Spawn calls

Extrys opened the pull request:
This PR introduces support for passing custom spawn data from the server to clients during the spawn process of a `NetworkObject`.  
This enables deterministic instantiation or contextual setup on the client without relying on RPCs or external sync steps.

Doesn't break or invalidate existing user code
the feature is fully optional and designed via an extension interface to preserve full backward compatibility.

## Changelog

- Added: `INetworkCustomSpawnDataReceiver` interface to allow reading custom spawn data on the client side.
- Changed: `CreateObjectMessage` now includes optional CustomSpawnData handling.
- Changed: Involved spawn methods now includes optional CustomSpawnData parameter.

## Testing and Documentation

- No tests have been added.

### Deprecated API
- No APIs were deprecated in this PR.


## Backport
- No backport required. This is a new feature for the current development version only.




Implementation example

```csharp
public class SyncedPrefabInstanceHandler : NetworkBehaviour, INetworkPrefabInstanceHandler, INetworkCustomSpawnDataReceiver
{
	public GameObject basePrefab;
	public NetworkObject basePrefabN;

	public GameObject[] prefabs;

	public override void OnNetworkSpawn()
	{
		NetworkManager.AddNetworkPrefab(basePrefab);
		NetworkManager.PrefabHandler.AddHandler(basePrefab, this);
	}
	public override void OnNetworkDespawn()
	{
		NetworkManager.PrefabHandler.RemoveHandler(basePrefab);
		NetworkManager.RemoveNetworkPrefab(basePrefab);
	}

       //See how i send custom data on the spawn
	public void Spawn(int metadata)
	{
		NetworkManager.SpawnManager.InstantiateAndSpawn(basePrefabN, customSpawnData: new byte[] { (byte)metadata});
	}

	int spawnData;
        // Implementing the INetworkCustomSpawnDataReceiver, i can catch this custom data just before the instantiate
	public void OnCustomSpawnDataReceived(byte[] customSpawnData)
	{
		Debug.Log("${nameof(SyncedPrefabInstanceHandler)}: Received custom spawn data: " + customSpawnData[0]);
		spawnData = customSpawnData[0];
	}
    
        //in the instantiate i use the custom spawn data allow looking for preinstantiated instances or spawning compatible arbitrary prefabs by id
	public NetworkObject Instantiate(ulong clientId, Vector3 position, Quaternion rotation)
	{
		return Instantiate(prefabs[spawnData], position, rotation).GetComponent<NetworkObject>();
	}

	public void Destroy(NetworkObject networkObject)
	{
		GameObject.Destroy(networkObject.gameObject);
	}
}```

Extrys committed:
Update package.json

Extrys committed:
Merge pull request #1 from Extrys/Extrys-patch-1

Update package.json

Extrys committed:
Allow send custom spawn data into the spawn command, no breaking changes

Extrys committed:
Removed work in progress

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3419) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3419) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3419) it.</sub>

Extrys committed:
Update package.json

EmandM commented:
Thank you for your contribution!

At this time, we would rather avoid adding parameters to the `Spawn()`, `SpawnWithOwnership()` and `SpawnPlayerObject()` methods. These are very high traffic methods and it is very easy for feature creep to happen in these places. This approach also introduces the potential for strange side effects around in scene placed Network Objects.

Could an approach like the `NetworkBehaviour.OnSynchronize()` method work for your use case? It might be easier if you could open an issue on this topic so we can discuss the use case more in-depth.

Extrys commented:
> Thank you for your contribution!
> 
> At this time, we would rather avoid adding parameters to the `Spawn()`, `SpawnWithOwnership()` and `SpawnPlayerObject()` methods. These are very high traffic methods and it is very easy for feature creep to happen in these places. This approach also introduces the potential for strange side effects around in scene placed Network Objects.
> 
> Could an approach like the `NetworkBehaviour.OnSynchronize()` method work for your use case? It might be easier if you could open an issue on this topic so we can discuss the use case more in-depth.

thanks for the feedback! i will open an issue regarding my use case

there is currently a forum post on unity i made for it, which is this
https://discussions.unity.com/t/allow-passing-spawn-metadata-to-instantiate-via-prefabhandler/1630903/2

i might copi this into an issue to see if there is a pleasing solution

Extrys commented:
> Hey, just a quick PR related note that you should probably rebase to target `develop-2.0.0` (our development branch for NGOv2.X) instead of `release/2.3.3`

Thanks for letting me know! I've just rebased the PR to target develop-2.0.0 as suggested 🙌

NoelStephensUnity commented:
> > Hey, just a quick PR related note that you should probably rebase to target `develop-2.0.0` (our development branch for NGOv2.X) instead of `release/2.3.3`
> 
> Thanks for letting me know! I've just rebased the PR to target `develop-2.0.0` as suggested 🙌 I'd just need to know if the approach I proposed above sounds viable, so I can start working on the updated implementation

Hi @Extrys,

Thank you for submitting your proposed changes. We always love community contributions...of course we might not always have enough time to review everything. Unfortunately, the change you are proposing is most likely not going to be approved for several reasons. The primary reason is that while your PR submission appears to cover the initial spawning of the NetworkObject it does not handle synchronizing late joining clients.

Example:

- A host starts a session.
  - Before any clients connect, it spawns 10 network prefabs using custom spawn information.
  - _Since there are no connected clients the data passed in via any `NetworkObject` spawn method will not be sent to any client._
- A client joins 5-10 seconds later.
  - The host begins the [Synchronization Process](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/10ef2eb49bfa2a1e50d91dd98b04ae80ed5790db/com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs#L2006).
    - This uses the [SceneEventMessage](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/10ef2eb49bfa2a1e50d91dd98b04ae80ed5790db/com.unity.netcode.gameobjects/Runtime/Messaging/Messages/SceneEventMessage.cs#L17) that invokes the [SceneEventData.Serialize](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/10ef2eb49bfa2a1e50d91dd98b04ae80ed5790db/com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs#L481) method.
      - Within the SceneEventData.Serialize method, it runs through all of the spawned NetworkObjects and invokes [NetworkObject.GetMessageSceneObject](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/10ef2eb49bfa2a1e50d91dd98b04ae80ed5790db/com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs#L3132) method.
        - This will serialize each NetworkObject based on the SceneObject structure populated, the NetworkVariables of each NetworkBehaviour, and any OnSynchronize custom serialization data per NetworkBehaviour.
        - This will no longer have access to the initial `customSpawnData` that was used on the host side to spawn the objects and you will still be confronted with the same issue.

Your implementation would only work under the condition of:

- The host and clients are already connected in a session and no NetworkObjects were spawned.
- The host then spawns NetworkObjects including the `customSpawnData`.
- The clients would receive the `customSpawnData`.
- Any late joining client would not receive the `customSpawnData`.

I [had posted a reply to your forum post here](https://discussions.unity.com/t/allow-passing-spawn-metadata-to-instantiate-via-prefabhandler/1630903/5) asking a bit more about why the ID assignment and more information as to why you needed this information during the instantiation as opposed to just using OnSynchronize as I had outlined.

If you could provide that information to me I might be able to assist and/or you could submit a bug report within the editor that includes your project and post that ticket here so I could look over your project and possibly provide some guidance on how you could achieve your goals without this type of modification... in the end the modification you are proposing would require storing the customData information in each instance permanently as well as change the existing public facing API (which both will be very difficult to get alignment for this without understanding your project/approach and determining if there is no other possible way to accomplish what you are trying to accomplish using the existing API).

Extrys committed:
Merge branch 'develop-2.0.0' into release/2.3.3

Extrys committed:
Symmetrical approach working

Extrys committed:
Clean diff and symetrical approach working for new InstantiationPayload

Extrys committed:
cleaning diff

Extrys commented:
Closing this PR in favor of a new, much cleaner and more aligned solution.
Addresses all feedback discussed here:
- Optional and non-breaking
- Fully aligns with NGO patterns (Serialize/Deserialize symmetry)
- No public API modifications
- Late-join and scene object support

Follow-up PR: [#XXXX (link to the new PR)](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3430)
